winter/schema/type_inspection.py

```diff
@@ -0,0 +1,195 @@
+import collections
+import datetime
+import decimal
+import enum
+import inspect
+import typing
+import uuid
+from collections import OrderedDict
+
+import dataclasses
+from drf_yasg import openapi
+from rest_framework.settings import api_settings as rest_settings
+
+from winter.type_utils import NoneType
+from winter.type_utils import UnionType
+from winter.type_utils import get_origin_type
+from winter.type_utils import is_optional
+
+TYPE_DECIMAL = openapi.TYPE_STRING if rest_settings.COERCE_DECIMAL_TO_STRING else openapi.TYPE_NUMBER
+TYPE_NONE = 'none'
+_resolvers_by_type: typing.Dict[typing.Type, typing.Tuple[typing.Callable, typing.Optional[typing.Callable]]] = {}
+
+
+def register(*types: typing.Tuple[typing.Type], checker=None):
+    def wrapper(func):
+        for type_ in types:
+            _resolvers_by_type[type_] = func, checker
+        return func
+
+    return wrapper
+
+
+class InspectorNotFound(Exception):
+
+    def __init__(self, hint_cls):
+        self.hint_cls = hint_cls
+
+    def __str__(self):
+        return f'Unknown type: {self.hint_cls}'
+
+
+@dataclasses.dataclass
+class TypeInfo:
+    type_: str
+    format_: typing.Optional[str] = None
+    child: typing.Optional['TypeInfo'] = None
+    nullable: bool = False
+    properties: typing.Dict[str, 'TypeInfo'] = dataclasses.field(default_factory=OrderedDict)
+    enum: list = None
+
+    def as_dict(self):
+        data = {
+            'type': self.type_,
+        }
+        if self.format_ is not None:
+            data['format'] = self.format_
+
+        if self.child is not None:
+            data['items'] = self.child.as_dict()
+
+        if self.nullable:
+            data['x-nullable'] = True
+
+        if self.enum is not None:
+            data['enum'] = self.enum
+
+        if self.properties:
+            data['properties'] = {key: value.as_dict() for key, value in self.properties.items()}
+
+        return data
+
+    def get_openapi_schema(self):
+        return openapi.Schema(**self.as_dict())
+
+
+# noinspection PyUnusedLocal
+@register(bool)
+def inspect_bool(hint_class) -> TypeInfo:
+    return TypeInfo(openapi.TYPE_BOOLEAN)
+
+
+# noinspection PyUnusedLocal
+@register(NoneType)
+def inspect_none(hint_class) -> TypeInfo:
+    return TypeInfo(TYPE_NONE)
+
+
+# noinspection PyUnusedLocal
+@register(int)
+def inspect_int(hint_class) -> TypeInfo:
+    return TypeInfo(openapi.TYPE_INTEGER)
+
+
+# noinspection PyUnusedLocal
+@register(str, bytes)
```

Comment:
here is str and bytes - I think it is not so bad because function must be same ( I think always)